pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -222,6 +225,17 @@ func (dswp *desiredStateOfWorldPopulator) findAndRemoveDeletedPods() {
 	for _, volumeToMount := range dswp.desiredStateOfWorld.GetVolumesToMount() {
 		pod, podExists := dswp.podManager.GetPodByUID(volumeToMount.Pod.UID)
 		if podExists {
+
+			// check if the attachability has changed for this volume
+			if volumeToMount.PluginIsAttachable &&
```

Comment:
Why do we need to delete the volume from volumemanager's DSOW as a special case? Is this merely to mark the fact that, volume is no longer attachable?

Reply:
Because our original `findAndRemoveDeletedPods` loop does not have the ability to detect if a volume plugin changes its attach-ability. Thus, if we dont add this check here, what happened is the reconciler will never happy and not able to bringup the pod because it always wait for that volume to be attached.

Reply:
can this cause a pod's volume to be unmounted if volume originally was attachable but somehow a CSIDriver deployment changed it to become unattachable?

Reply:
But `IsAttachableVolume` check depends on informers and cache being in sync (in other words network), so it is not just a local state. For example - a CSI plugin that was attachable previously can be considered non-attachable if informers are being resynced or CSIDriver object can't be found (because of network error).  In which case kubelet would proceed to unmount pod's volume and can cause data loss and data corruption.

Reply:
Hmm, yeah. That is a solid concern. But I guess the question is then how do we decide if it is a network issue or is the attachability changed for a CSIDriver. Also, I am new to this so I might be wrong, but isn't informers and cache also a local state? I am wondering how other components handle situation like this. @saad-ali any concerns or thoughts here?

Reply:
At very minimum - `IsAttachableVolume` check is flawed as means of determining whether volume should be removed from desired state of the world. If volume was previously attachable  but anything fails in `FindAttachablePluginBySpec`, we assume volume is not attachable anymore. Whereas this should be other way around - if and only if - `FindAttachablePluginBySpec` returns nil plugin but no error, we can safely assume that plugin is indeed not attachable.  The burden of verifying whether plugin is attachable or not is much higher in this case.

But I also think that - we are making this fix in wrong place IMO. Deleting the pod from desired state as I understand will cause all previous assumptions to be invalidated and hence pod can be started.  But can we not fix this in `GenerateMountXXX` function, so as if a plugin changes its attchability then mount operation does not wait for volume to become attached? is something I missed?

Reply:
I agree with Hemant that an unmount/mount while the container is still running is undesirable and could cause data corruption/loss.

It may be useful to think about the different scenarios we could run into when a CSIDriver changes from attachable to non-attachable:

1. Before AD controller adds it to DSW
2. After AD controller adds it to DSW and before it creates the VolumeAttachment object
3. After AD controller creates the VolumeAttachment object and kubelet is stuck waiting (because there's no sidecar running)
4. Volume is already mounted and pod is already running (because previously the sidecar was running before the driver got upgraded)

I think the behavior we want is:
- Be able to resolve scenarios 2 + 3
- Delete the VolumeAttachment object
- Clean up Node.Status.VolumesAttached
- Clean up any internal state
- All the while, not disturbing any mounts

Reply:
Thanks Hemant and Michelle for your feedback!

Re Michelle:
I also agree that unmount/mount while container is running is undesirable. But this might only happen if there is already a pod running using the volume and we change the volume from attachable to non-attachable. In that case, it is the user's choice to do the flip.

The problem with the scenarios you mentioned is that we not only need to care about A/D controller, but we also need to care about volumemanager which resides in the kubelet side. In the PR that I already made, we have taken care of the A/D side, so let's break down the scenarios you mentioned for A/D (The result is based on the PR I have):
1. There is no problem on this, because if the attach-ability change happens before AD add it to DSW, then it will never add it to DSW.
2. In my fix, there is a race between the VolumeAttachment creation and the removal of the volume from DSW. If VolumeAttachment creation happens first, then what gonna happen is we will remove the volume from DSW, and a detach operation will be triggered and then the VolumeAttachment will be deleted. If the removal happens first, then VolumeAttachment will not be created which is also good.
3. As mentioned in 2, we will remove the volume from DSW, and a detach operation will kick in. The detach will either timeout or get an update from nodeStatus to inform it that the volume is no long in the attachableVolume list. Both cases the VolumeAttachment will then be deleted.
4. The only problematic scenario is this one: so what will happen is that a detach operation will be triggered and there are multiple different outcomes depending on the driver. There might be cases where even the CSIDriver said it does not support attach/detach, it can still perform a detach operation which will make the pod lost its access to the volume anymore. There might be cases where the driver does not support detach operation so the detach for it is a no-op. Besides, we will perform a unmount/mount for the pod.

So I guess we have satisfied the behavior you want except `All the while, not disturbing any mounts`. But then the question is, if the pod is already running with the CSIDriver, why do we change the driver's attachability? The only possible scenario is we upgraded the CSIDriver to a version where it does not support attach/detach anymore, which IMO it should be deployed with a different CSIDriver name. So it is really the user's call if they decided to meddling the driver's attachability if there is already running pods using that driver.


Re: Hemant
I agree we need to loose the isAttachableVolume check so that it only states a volume is nonattachable when no err and it return nil.

Now let's talk about why I am deleting the pod from volume, it is true that it will invalidate the assumptions we made, but that comes with a good reason: we changed the volume attachability. Thus, a lot of assumptions we made in the previous run is wrong. Here is a detailed explanation:
The DSW in volumemanager keeps the records of all Volumes it needs to mount. However, it keeps a record of the volume attach-ability when it adds the volume into its list here: https://github.com/kubernetes/kubernetes/blob/04362870ad9f9812f487ab8f89a40f6bf2daa588/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L169
So in the beginning I thought maybe we can just flip the boolean so that it can proceed. But then I found out I cannot do so because the `VolumeUniqueName` which is the primary key for the volumeToMount list in the DSW is generated based on its attach-ability:
https://github.com/kubernetes/kubernetes/blob/04362870ad9f9812f487ab8f89a40f6bf2daa588/pkg/kubelet/volumemanager/cache/desired_state_of_world.go#L234
So there might be cases if we deployed the CSIDriver, the whole volume unique name for that volume will be changed. That can be problematic because the original volumeToMount in DSW will never be satisfied and thus the pod can never be bringup.

I could be wrong because the logic here is quite complicated but it is great to discuss in details for this.

Reply:
To clarify what "driver is not attachable" means, there's a couple of scenarios:
1. CSI driver does not implement ControllerPublish, but is running the csi-attacher in the simple mode where all it does is set the VolumeAttachment.Status.attached = true without calling to the driver. This was what drivers had to do before we introduced the CSIDriver.Spec,AttachedRequired field.
2. CSI driver does not implement ControllerPublish, is not running csi-attacher, and creates a CSIDriver object and sets AttachedRequired = false.

Upgrading from 1->2 will hit this issue and I expect any non-attachable driver that was deployed before CSIDriver.AttachedRequired existed can have existing volumes that are mounted. I don't think we can leave it to the users to handle this case. The admin managing the CSI driver is different than the user running Pods, and in managed distributions that bundle CSI drivers, the upgrade of the CSI driver is potentially out of the user's control.

2) will hit this issue if the CSIDriver object is deployed later. Potentially this could be mitigated by deployed the csi driver well before any workloads can run, but that may not always be controllable.

Reply:
Hmm, yeah that makes sense. I am trying to find out if we can update the internal state of DSW directly instead delete and create it.

Reply:
2 concerns stand out to me: 1) potential for detach or unmount of in-use volume, and 2) assuming default value on failure to fetch CSIDriver object.

--1--

With @Jiawei0227's current proposal, in this case, the A/D controller would delete the VolumeAttachment object because it depends on Node `volumesInUse` list for safe detach. When a volume switches from attachable to non-attachable, the volume is removed from `volumesInUse` by the kubelet as soon as possible, meaning A/D controller could do unsafe detach. This is problematic because of the use case @msau42 pointed out:

>>> Volume is already mounted and pod is already running (because previously the sidecar was running before the driver got upgraded)
>> 
>> The only problematic scenario is this one: so what will happen is that a detach operation will be triggered and there are multiple different outcomes depending on the driver. There might be cases where even the CSIDriver said it does not support attach/detach, it can still perform a detach operation which will make the pod lost its access to the volume anymore. There might be cases where the driver does not support detach operation so the detach for it is a no-op. Besides, we will perform a unmount/mount for the pod.
>
> Upgrading from 1->2 will hit this issue and I expect any non-attachable driver that was deployed before CSIDriver.AttachedRequired existed can have existing volumes that are mounted.

This^^ is a valid use case, and we should handle it gracefully. 
To handle it correctly, I don't think any further changes are needed on the a/d controller side, since it already depends on the kubelet to tell it when it is safe to detach -- the problem is on the kubelet side, with the new logic. the volume is removed from `volumeInUse` as soon as it switches from attachable to non-attachable. So would it be sufficient to add a condition on the kubelet volume manager side that does not remove from DSW if the pod is already running (@Jiawei0227 was considering this when we chatted yesterday)?

--2--

>> At very minimum - IsAttachableVolume check is flawed as means of determining whether volume should be removed from desired state of the world. If volume was previously attachable but anything fails in FindAttachablePluginBySpec, we assume volume is not attachable anymore. Whereas this should be other way around - if and only if - FindAttachablePluginBySpec returns nil plugin but no error, we can safely assume that plugin is indeed not attachable. The burden of verifying whether plugin is attachable or not is much higher in this case.
> 
> Hmm, yeah. That is a solid concern. But I guess the question is then how do we decide if it is a network issue or is the attachability changed for a CSIDriver. Also, I am new to this so I might be wrong, but isn't informers and cache also a local state? I am wondering how other components handle situation like this. @saad-ali any concerns or thoughts here?

This is also a valid concern. Ideally a failure should mean both A/d controller and kubelet volume manager continue to assume prior state about attachability. Can we surface errors, and compare them to assumed attachability state? If that is really difficult to do, we can consider sane defaults, but we'll need to think through the implications of that.

Reply:
I met with @Jiawei0227 to discuss this.

Regarding issue number 1 above in my comment, after discussing with @Jiawei0227 I don't think it is a problem worth fixing. Because, In the case Michelle pointed out going from attachable (CSI Attacher with no-op attach) to non-attachable (CSI Driver object saying skip attach), even if detach is inadvertently called, because CSI attacher was a no-op, nothing bad should happen (no risk of data corruption on a no-op detach). The only case this would be a concern would be if user creates a CSIDriver object saying skip attach for a driver that implements a real attach/detach -- that does not seem like a realistic scenario.

Regarding issue number 2 above in my comment, after discussing with @Jiawei0227 he agreed to have the attachability check surface any error to his attachability checks in both A/d controller and kubelet volume manager, and if there is an error, have the checks continue to assume prior state about attachability (don't assumes state change unless explicitly confirmed). He will update his code to `FindAttachablePluginBySpec`.

Finally @Jiawei0227 pointed out @gnufied's main concern, which was to prevent rapid unmount/mount of volumes in use that could happen with the existing proposal. I agree this is a major issue. To fix it, @Jiawei0227 proposed instead of creating/deleting volumes in volume manager DSW, he will update the attachability state for the volume instead.

Reply:
After discussing with @saad-ali 

We we decide to go with an update mechanism instead of a delete/create machoism for volumes in DSW of volume manager. This way we will not trigger an unmount/mount for existing running pods and thus bring no harms. Also, I have adopted what @gnufied suggested which is to use the FindAttachablePluginBySpec and only state a plugin is non-attachable if no err and nil volume plugin return.